File: app/src/main/java/jp/juggler/fadownloader/Pref.java
package jp.juggler.fadownloader;

import android.content.Context;
import android.content.SharedPreferences;
import android.text.TextUtils;

public class Pref{

	public static SharedPreferences pref( Context context ){
		return context.getSharedPreferences( "app_pref", Context.MODE_PRIVATE );
	}

	public static final int TARGET_TYPE_FLASHAIR_AP = 0;
	public static final int TARGET_TYPE_FLASHAIR_STA = 1;
	public static final int TARGET_TYPE_PENTAX_KP = 2;

	public static String loadTargetUrl( SharedPreferences pref, int target_type ){
		switch( target_type ){
		default:
		case TARGET_TYPE_FLASHAIR_AP:
			return pref.getString( Pref.UI_TARGET_URL_0, "http://flashair/"  );

		case TARGET_TYPE_FLASHAIR_STA:
			return pref.getString( Pref.UI_TARGET_URL_1, "http://flashair/" );

		case TARGET_TYPE_PENTAX_KP:
			return pref.getString( Pref.UI_TARGET_URL_2, "http://192.168.0.1/" );
		}
	}

	public static void saveTargetUrl( SharedPreferences.Editor edit, int target_type, String value ){
		switch( target_type ){
		case TARGET_TYPE_FLASHAIR_AP:
			edit.putString( Pref.UI_TARGET_URL_0, value );
			break;

		case TARGET_TYPE_FLASHAIR_STA:
			edit.putString( Pref.UI_TARGET_URL_1, value );
			break;

		case TARGET_TYPE_PENTAX_KP:
			edit.putString( Pref.UI_TARGET_URL_2, value );
			break;
		}
	}

	// UI画面に表示されている情報の永続化
	public static final String UI_REPEAT = "ui_repeat";
	public static final String UI_LAST_PAGE = "ui_last_page";
	public static final String UI_TARGET_TYPE = "ui_target_type";
	public static final String UI_TARGET_URL_0 = "ui_flashair_url"; // 歴史的な理由でキー名が特別
	public static final String UI_TARGET_URL_1 = "ui_target_url_1";
	public static final String UI_TARGET_URL_2 = "ui_target_url_2";
	public static final String UI_FOLDER_URI = "ui_folder_uri";
	public static final String UI_INTERVAL = "ui_interval";
	public static final String UI_FILE_TYPE = "ui_file_type";
	public static final String UI_LOCATION_MODE = "ui_location_mode";
	public static final String UI_LOCATION_INTERVAL_DESIRED = "ui_location_interval_desired";
	public static final String UI_LOCATION_INTERVAL_MIN = "ui_location_interval_min";
	public static final String UI_FORCE_WIFI = "ui_force_wifi";
	public static final String UI_SSID = "ui_ssid";
	public static final String UI_THUMBNAIL_AUTO_ROTATE = "ui_thumbnail_auto_rotate";
	public static final String UI_COPY_BEFORE_VIEW_SEND = "ui_copy_before_view_send";

	public static final boolean DEFAULT_THUMBNAIL_AUTO_ROTATE = true;

	public static void initialize( Context context ){
		SharedPreferences pref = pref( context );

		SharedPreferences.Editor e = pref.edit();
		boolean bChanged = false;
		String sv;
		int iv;


		//
		sv = pref.getString( Pref.UI_INTERVAL, null );
		if( TextUtils.isEmpty( sv ) ){
			bChanged = true;
			e.putString( Pref.UI_INTERVAL, "30" );
		}
		//
		sv = pref.getString( Pref.UI_FILE_TYPE, null );
		if( TextUtils.isEmpty( sv ) ){
			bChanged = true;
			e.putString( Pref.UI_FILE_TYPE, ".jp*" );
		}
		//
		iv = pref.getInt( Pref.UI_LOCATION_MODE, - 1 );
		if( iv < 0 || iv > LocationTracker.LOCATION_HIGH_ACCURACY ){
			bChanged = true;
			e.putInt( Pref.UI_LOCATION_MODE, LocationTracker.DEFAULT_MODE );
		}
		//
		sv = pref.getString( Pref.UI_LOCATION_INTERVAL_DESIRED, null );
		if( TextUtils.isEmpty( sv ) ){
			bChanged = true;
			e.putString( Pref.UI_LOCATION_INTERVAL_DESIRED
				, Long.toString( LocationTracker.DEFAULT_INTERVAL_DESIRED / 1000L ) );
		}
		//
		sv = pref.getString( Pref.UI_LOCATION_INTERVAL_MIN, null );
		if( TextUtils.isEmpty( sv ) ){
			bChanged = true;
			e.putString( Pref.UI_LOCATION_INTERVAL_MIN
				, Long.toString( LocationTracker.DEFAULT_INTERVAL_MIN / 1000L ) );
		}
		//
		if( bChanged ) e.apply();
	}

	// 最後に押したボタン
	public static final String LAST_MODE_UPDATE = "last_mode_update";
	public static final String LAST_MODE = "last_mode";
	public static final int LAST_MODE_STOP = 0;
	public static final int LAST_MODE_ONCE = 1;
	public static final int LAST_MODE_REPEAT = 2;

	// 最後にWorkerを手動開始した時の設定
	public static final String WORKER_REPEAT = "worker_repeat";
	public static final String WORKER_TARGET_TYPE = "worker_target_type";
	public static final String WORKER_FLASHAIR_URL = "worker_flashair_url";
	public static final String WORKER_FOLDER_URI = "worker_folder_uri";
	public static final String WORKER_INTERVAL = "worker_interval";
	public static final String WORKER_FILE_TYPE = "worker_file_type";
	public static final String WORKER_LOCATION_INTERVAL_DESIRED = "worker_location_interval_desired";
	public static final String WORKER_LOCATION_INTERVAL_MIN = "worker_location_interval_min";
	public static final String WORKER_LOCATION_MODE = "worker_location_mode";
	public static final String WORKER_FORCE_WIFI = "worker_force_wifi";
	public static final String WORKER_SSID = "worker_ssid";

	// ファイルスキャンが完了した時刻
	public static final String LAST_IDLE_START = "last_idle_start";
	public static final String FLASHAIR_UPDATE_STATUS_OLD = "flashair_update_status_old";

	public static final String REMOVE_AD_PURCHASED = "remove_ad_purchased";

}

File: app/src/main/java/jp/juggler/fadownloader/Pref.kt
package jp.juggler.fadownloader

import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import jp.juggler.fadownloader.tracker.LocationTracker

open class BasePref(val key : String)

fun SharedPreferences.Editor.remove(bp : BasePref) : SharedPreferences.Editor =
	this.remove(bp.key)

class BooleanPref(
	key : String,
	private val defVal : Boolean
) : BasePref(key) {
	
	operator fun invoke(pref : SharedPreferences) = pref.getBoolean(key, defVal)
	operator fun invoke(intent : Intent) = intent.getBooleanExtra(key, defVal)
}

fun SharedPreferences.Editor.put(bp : BooleanPref, v : Boolean) : SharedPreferences.Editor =
	this.putBoolean(bp.key, v)

class IntPref(
	key : String,
	val defVal : Int
) : BasePref(key) {
	
	operator fun invoke(pref : SharedPreferences) = pref.getInt(key, defVal)
	operator fun invoke(intent : Intent) = intent.getIntExtra(key, defVal)
}

fun SharedPreferences.Editor.put(bp : IntPref, v : Int) : SharedPreferences.Editor =
	this.putInt(bp.key, v)

class LongPref(
	key : String,
	private val defVal : Long
) : BasePref(key) {
	
	operator fun invoke(pref : SharedPreferences) = pref.getLong(key, defVal)
}

fun SharedPreferences.Editor.put(bp : LongPref, v : Long) : SharedPreferences.Editor =
	this.putLong(bp.key, v)

class StringPref(
	key : String,
	val defVal : String
) : BasePref(key) {
	
	operator fun invoke(pref : SharedPreferences) : String = pref.getString(key, defVal) ?: defVal
	operator fun invoke(intent : Intent) : String = intent.getStringExtra(key) ?: defVal
	
	//	fun getInt(pref : SharedPreferences) = try {
	//		invoke(pref).trim().toInt()
	//	} catch(ex : Throwable) {
	//		defVal.toInt()
	//	}
	
	fun getInt(intent : Intent) = try {
		invoke(intent).trim().toInt()
	} catch(ex : Throwable) {
		defVal.toInt()
	}
	
	fun getIntOrNull(pref : SharedPreferences) = try {
		invoke(pref).trim().toInt()
	} catch(ex : Throwable) {
		null
	}
	
	//	fun getIntOrNull(intent:Intent) = try {
	//		invoke(intent).trim().toInt()
	//	} catch(ex : Throwable) {
	//		null
	//	}
}

fun SharedPreferences.Editor.put(p : StringPref, v : String) : SharedPreferences.Editor =
	this.putString(p.key, v)

fun Intent.put(pref : SharedPreferences, p : BooleanPref) {
	putExtra(p.key, p(pref))
}

fun Intent.put(pref : SharedPreferences, p : StringPref) {
	putExtra(p.key, p(pref))
}

fun Intent.put(pref : SharedPreferences, p : IntPref) {
	putExtra(p.key, p(pref))
}

fun Intent.put(v : String, p : StringPref) {
	putExtra(p.key, v)
}

fun Intent.put(v : Int, p : IntPref) {
	putExtra(p.key, v)
}

object Pref {
	fun pref(context : Context) : SharedPreferences {
		return context.getSharedPreferences("app_pref", Context.MODE_PRIVATE)
	}
	
	const val TARGET_TYPE_FLASHAIR_AP = 0
	const val TARGET_TYPE_FLASHAIR_STA = 1
	const val TARGET_TYPE_PENTAX_KP = 2
	const val TARGET_TYPE_PQI_AIR_CARD = 3
	const val TARGET_TYPE_PQI_AIR_CARD_TETHER = 4
	
	const val LAST_MODE_STOP = 0
	const val LAST_MODE_ONCE = 1
	const val LAST_MODE_REPEAT = 2
	
	private const val DEFAULT_WIFI_AP_CHANGE_INTERVAL = 5000L
	private const val DEFAULT_WIFI_SCAN_INTERVAL = 10000L
	private const val DEFAULT_TETHER_SPRAY_INTERVAL = 3000L
	private const val DEFAULT_TETHER_TEST_CONNECTION_TIMEOUT = 15000L
	
	private const val DEFAULT_LOCATION_MODE = LocationTracker.NO_LOCATION_UPDATE
	private const val DEFAULT_LOCATION_INTERVAL_DESIRED = 1000L * 3600
	private const val DEFAULT_LOCATION_INTERVAL_MIN = 1000L * 300
	
	
	// UI画面に表示されている情報の永続化
	val uiRepeat = BooleanPref("ui_repeat", false)
	val uiForceWifi = BooleanPref("ui_force_wifi", false)
	val uiAutoRotateThumbnail = BooleanPref("ui_thumbnail_auto_rotate", true)
	val uiCopyBeforeSend = BooleanPref("ui_copy_before_view_send", false)
	val uiProtectedOnly = BooleanPref("ui_protected_only", false)
	val uiSkipAlreadyDownload = BooleanPref("ui_skip_already_download", false)
	val uiStopWhenTetheringOff = BooleanPref("uiStopWhenTetheringOff", false)
	
	val uiLastPage = IntPref("ui_last_page", 0)
	val uiTargetType = IntPref("ui_target_type", - 1)
	val uiLocationMode = IntPref("ui_location_mode", DEFAULT_LOCATION_MODE)
	
	val uiFolderUri = StringPref("ui_folder_uri", "")
	val uiFileType = StringPref("ui_file_type", ".jp*")
	val uiSsid = StringPref("ui_ssid", "")
	
	val uiInterval = StringPref(
		"ui_interval",
		"30" // unit : second
	)
	
	val uiTetherSprayInterval = StringPref(
		"uiTetherSprayInterval",
		(DEFAULT_TETHER_SPRAY_INTERVAL / 1000L).toString()
	)
	
	val uiTetherTestConnectionTimeout = StringPref(
		"uiTetherTestConnectionTimeout",
		(DEFAULT_TETHER_TEST_CONNECTION_TIMEOUT / 1000L).toString()
	)
	val uiWifiChangeApInterval = StringPref(
		"uiWifiChangeApInterval",
		(DEFAULT_WIFI_AP_CHANGE_INTERVAL / 1000L).toString()
	)
	val uiWifiScanInterval = StringPref(
		"uiWifiScanInterval",
		(DEFAULT_WIFI_SCAN_INTERVAL / 1000L).toString()
	)
	
	val uiLocationIntervalDesired = StringPref(
		"ui_location_interval_desired",
		(DEFAULT_LOCATION_INTERVAL_DESIRED / 1000L).toString()
	)
	
	val uiLocationIntervalMin = StringPref(
		"ui_location_interval_min",
		(DEFAULT_LOCATION_INTERVAL_MIN / 1000L).toString()
	)
	
	//////////////////////////////////////////////////////////////////////
	
	private val uiTargetUrlFlashAirAp = StringPref(
		"ui_flashair_url", // 歴史的な理由でキー名が特別
		"http://flashair/"
	)
	private val uiTargetUrlFlashAirSta = StringPref(
		"ui_target_url_1",
		"http://AutoDetect/"
	)
	private val uiTargetUrlPentaxKp = StringPref(
		"ui_target_url_2",
		"http://192.168.0.1/"
	)
	private val uiTargetUrlPentaxPqiAirCard = StringPref(
		"ui_target_url_pqi_air_card",
		"http://192.168.1.1/"
	)
	private val uiTargetUrlPentaxPqiAirCardTether = StringPref(
		"ui_target_url_pqi_air_card_tether",
		"http://AutoDetect/"
	)
	
	private fun uiTargetUrl(target_type : Int) = when(target_type) {
		TARGET_TYPE_FLASHAIR_AP -> uiTargetUrlFlashAirAp
		TARGET_TYPE_FLASHAIR_STA -> uiTargetUrlFlashAirSta
		TARGET_TYPE_PENTAX_KP -> uiTargetUrlPentaxKp
		TARGET_TYPE_PQI_AIR_CARD -> uiTargetUrlPentaxPqiAirCard
		TARGET_TYPE_PQI_AIR_CARD_TETHER -> uiTargetUrlPentaxPqiAirCardTether
		else -> uiTargetUrlFlashAirAp
	}
	
	fun loadTargetUrl(pref : SharedPreferences, target_type : Int) =
		uiTargetUrl(target_type)(pref)
	
	fun saveTargetUrl(edit : SharedPreferences.Editor, target_type : Int, value : String) =
		edit.put(uiTargetUrl(target_type), value)
	
	//////////////////////////////////////////////////////////////////////
	
	// 最後にWorkerを手動開始した時の設定
	
	val workerRepeat = BooleanPref("worker_repeat", false)
	val workerForceWifi = BooleanPref("worker_force_wifi", false)
	val workerProtectedOnly = BooleanPref("worker_protected_only", false)
	val workerSkipAlreadyDownload = BooleanPref("worker_skip_already_download", false)
	val workerStopWhenTetheringOff = BooleanPref("workerStopWhenTetheringOff", false)
	
	val workerTargetUrl = StringPref("worker_flashair_url", "")
	val workerFolderUri = StringPref("worker_folder_uri", "")
	val workerFileType = StringPref("worker_file_type", ".jp*")
	val workerSsid = StringPref("worker_ssid", "")
	
	val workerTargetType = IntPref("worker_target_type", - 1)
	val workerLocationMode = IntPref("worker_location_mode", DEFAULT_LOCATION_MODE)
	
	val workerInterval = IntPref(
		"worker_interval",
		86400 // unit : second
	)
	
	val workerLocationIntervalDesired = LongPref(
		"worker_location_interval_desired",
		DEFAULT_LOCATION_INTERVAL_DESIRED
	)
	
	val workerLocationIntervalMin = LongPref(
		"worker_location_interval_min",
		DEFAULT_LOCATION_INTERVAL_MIN
	)
	
	val workerTetherSprayInterval = LongPref(
		"workerTetherSprayInterval",
		DEFAULT_TETHER_SPRAY_INTERVAL
	)
	
	val workerTetherTestConnectionTimeout = LongPref(
		"workerTetherTestConnectionTimeout",
		DEFAULT_TETHER_TEST_CONNECTION_TIMEOUT
	)
	
	val workerWifiChangeApInterval = LongPref(
		"workerWifiChangeApInterval",
		DEFAULT_WIFI_AP_CHANGE_INTERVAL
	)
	
	val workerWifiScanInterval = LongPref(
		"workerWifiScanInterval",
		DEFAULT_WIFI_SCAN_INTERVAL
	)
	
	//////////////////////////////////////////////////////////////////////
	
	// 最後に押した動作ボタンとその時刻
	val lastMode = IntPref("last_mode", LAST_MODE_STOP)
	val lastModeUpdate = LongPref("last_mode_update", 0L)
	
	// ファイルスキャンが完了した時刻
	val lastIdleStart = LongPref("last_idle_start", 0L)
	val flashAirUpdateStatusOld = LongPref("flashair_update_status_old", - 1L)
	
	// ダウンロード完了通知に表示する数字
	val downloadCompleteCount = LongPref("download_complete_count", 0L)
	val downloadCompleteCountHidden = LongPref("download_complete_count_hidden", 0L)
	
	val purchasedRemoveAd = BooleanPref("remove_ad_purchased", false)
	
	//////////////////////////////////////////////////////////////////////
	
	fun initialize(context : Context) {
		val pref = pref(context)
		
		var bChanged = false
		
		val e = pref.edit()
		
		// 空文字列ならデフォルト値に設定しなおす
		for(sp in arrayOf(
			uiInterval,
			uiFileType,
			uiLocationIntervalDesired,
			uiLocationIntervalMin
		)) {
			if(sp(pref).trim().isEmpty()) {
				bChanged = true
				e.put(sp, sp.defVal)
			}
		}
		
		// 範囲外の値ならデフォルト値に設定しなおす
		if(uiLocationMode(pref) !in 0 .. LocationTracker.LOCATION_HIGH_ACCURACY) {
			bChanged = true
			e.put(uiLocationMode, uiLocationMode.defVal)
		}
		
		if(bChanged) e.apply()
	}
}
